### PR TITLE
Fix undo/redo merge bugs

### LINKED ISSUE #883: Pushing mergeable Commands after undo doesn't delete rest of stack
Given a `CommandStack` in normal operation (ie. not building a macro) where you are doing a sequence of steps of the form:

- `push` some commands
- `undo` one or more commands
- `push` a command

then all commands after the current command are deleted from the stack here:
https://github.com/enthought/pyface/blob/05eefd485fa3831d39fdf4156f24bdc0d8e5140f/pyface/undo/command_stack.py#L211

However if the command which is pushed happens to be mergeable, then the remainder of the stack is not cleared.  It should be.

For comparison, the `UndoHandler` of TraitsUI does do this correctly.

### LINKED ISSUE #884: Pushing mergeable Commands at clean point shouldn't merge
Given a `CommandStack` in normal operation (ie. not building a macro) where you are doing a sequence of steps of the form:

- `push` a command (or otherwise have the current command be mergeable)
- mark the stack as `clean` at that point (eg. user saves at that point)
- `push` a command that can be merged with the current command

This should not be merged, as then the "clean" state would be modified which might prevent operations that are contingent on being in a "dirty" state (eg. "Save" a document).

This is not state that the `Command` can see, so it can't change its behaviour based on this, so the `CommandStack` is the right place to handle it.

### PR DESCRIPTION
**Pull request by corranwebster**
Add tests for undo/redo and fix #883 and fix #884 

Not much more.

### CODE REVIEW

File: pyface/undo/command_stack.py [hunk 1]
```diff
@@ -201,6 +201,9 @@ def push(self, command):
 
         # Do nothing more if the command was merged.
```
Comment (rahulporuri) [inline pyface/undo/command_stack.py:202]: this comment needs to be updated - or removed - given the recent changes.

File: pyface/undo/tests/test_command_stack.py [hunk 2]
```diff
@@ -31,21 +34,92 @@ def test_empty_command_stack(self):
...
+
+    def test_n_merge_command_pushed(self):
+        n = 4
+        with self.assert_n_commands_pushed(self.stack, 1):
+            self.command = MergeableCommand()
+            self.stack.push(self.command)
+            for i in range(n):
+                command = MergeableCommand()
+                with self.assertTraitChanges(self.stack.undo_manager,
+                                         'stack_updated', count=1):
```
Comment (rahulporuri) [inline pyface/undo/tests/test_command_stack.py:87]: nit-picky suggestion

```suggestion
                with self.assertTraitChanges(self.stack.undo_manager,
                                             'stack_updated', count=1):
```

**Review by rahulporuri — APPROVED** [review]
LGTM with two minor comments.

**Comment by rahulporuri** [comment]
Also, given that the 7.3.0 release is imminent, do we need to backport this PR to the 7.2 maintenance branch for a 7.2.1 bug fix release? We could backport the changes purely in principle, in case someone requests a 7.2.1 specifically and can't use 7.3.0.

**Comment by corranwebster** [comment]
If 7.3 is about to come out, then we don't need to do the backport, but maybe keep the flag on the PR in case for some reason we later need a 7.2.1 release and then we can easily work out what needs to be backported.

**Pull request merged**